### PR TITLE
feat: add typings for <Button>

### PR DESCRIPTION
**Pull request by bradenmacdonald**
## Description

This adds typescript types for `<Button>` as well as `<ButtonGroup>` and `<ButtonToolbar>`.

I tested it against various usages of `<Button>` in frontend-app-course authoring, as well as all the usages in this repo itself, and couldn't find any issues.

### Deploy Preview

https://deploy-preview-3080--paragon-openedx.netlify.app

## Merge Checklist

* [x] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Netlify deploy preview, if applicable.
* [x] Does your change adhere to the documented [style conventions](https://github.com/openedx/paragon/blob/master/docs/decisions/0012-css-styling-conventions)?
* [x] Do any prop types have missing descriptions in the Props API tables in the documentation site (check deploy preview)?
* [x] Were your changes tested using all available themes (see theme switcher in the header of the deploy preview, under the "Settings" icon)?
* [x] Were your changes tested in the `example` app?
* [x] Is there adequate test coverage for your changes?
* [x] Consider whether this change needs to reviewed/QA'ed for accessibility (a11y). If so, please request an a11y review for the PR in the [#wg-paragon](https://openedx.slack.com/archives/C02NR285KV4) Open edX Slack channel.

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@openedx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @bradenmacdonald! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: src/Button/Button.test.tsx [hunk 1]
```diff
@@ -94,9 +94,21 @@ describe('<Button />', () => {
       });
 
       test('test button as hyperlink', () => {
-        render(<Button as={Hyperlink} destination="https://www.poop.com/💩">Button</Button>);
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        const ref = (_current: HTMLAnchorElement) => {}; // Check typing of a ref - should not show type errors.
+        render(<Button as={Hyperlink} ref={ref} destination="https://www.poop.com/💩">Button</Button>);
```
Comment (bradenmacdonald) [inline src/Button/Button.test.tsx:99]: This shows an example of how passing `as={Hyperlink}` changes the types of various other props like `ref` to match.

File: src/Button/Button.test.tsx [hunk 2]
```diff
@@ -30,7 +30,7 @@ describe('<Button />', () => {
 
     it('renders with props iconAfter and size', () => {
       const tree = renderer.create((
-        <Button iconAfter={Close} size="md">Button</Button>
+        <Button iconAfter={Close} size="sm">Button</Button>
```
Comment (bradenmacdonald) [inline src/Button/Button.test.tsx:33]: I just changed this to `sm` for some variation - too many of the tests were just using the same `md` size.

File: src/Button/index.tsx [hunk 3]
```diff
@@ -1,32 +1,57 @@
...
-import PropTypes from 'prop-types';
+import PropTypes, { type Requireable } from 'prop-types';
 import classNames from 'classnames';
-import BaseButton from 'react-bootstrap/Button';
-import BaseButtonGroup from 'react-bootstrap/ButtonGroup';
-import BaseButtonToolbar from 'react-bootstrap/ButtonToolbar';
+import BaseButton, { type ButtonProps as BaseButtonProps } from 'react-bootstrap/Button';
+import BaseButtonGroup, { type ButtonGroupProps as BaseButtonGroupProps } from 'react-bootstrap/ButtonGroup';
+import BaseButtonToolbar, { type ButtonToolbarProps } from 'react-bootstrap/ButtonToolbar';
+import { type BsPrefixRefForwardingComponent as ComponentWithAsProp } from 'react-bootstrap/esm/helpers';
```
Comment (adamstankiewicz) [inline src/Button/index.tsx:7]: [question] Given this pattern of importing `BsPrefixRefForwardingComponent` (and `BsPrefixProps`, see https://github.com/openedx/paragon/pull/3082/), I wonder if there's a more common/global location to define these types when renaming them so each individual component doesn't need to do the same re-naming repeatedly.

For example, `Hyperlink` (in https://github.com/openedx/paragon/pull/3082/) also relies on this pattern so it'd be nice to only rename it in one place.

Comment (bradenmacdonald) [reply src/Button/index.tsx:7]: Yes, that's a good idea and I'm happy to move it somewhere shared. Do you have a place in mind or want me to make something up?

Comment (brian-smith-tcril) [reply src/Button/index.tsx:7]: First thing that came to my mind was to rename https://github.com/openedx/paragon/tree/master/src/utils/propTypes to `src/utils/types` and then put it in there, but I'm very open to any other ideas.

**Review by brian-smith-tcril — APPROVED** [review]
This is wonderful! Thank you!

**Comment by brian-smith-tcril** [comment]
I believe codecov is stuck, I'm going to merge this now instead of waiting for that check

image

**Comment by openedx-webhooks** [comment]
@bradenmacdonald 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/paragon&entry.867055334=https://github.com/openedx/paragon/pull/3080&entry.1484655318=https://github.com/bradenmacdonald&entry.752974735=2024-05-22+18:19&entry.1917517419=2024-05-31+17:26&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by bradenmacdonald** [comment]
Thanks @brian-smith-tcril !

**Pull request merged**